gwpy/utils/lal.py

```diff
@@ -144,6 +157,33 @@ def find_typed_function(pytype, prefix, suffix, module=lal):
     return getattr(module, '{0}{1}{2}'.format(prefix, laltype, suffix))
 
 
+def from_lal_dtype(lalts):
+    """Convert the LAL time-series data type into a numpy data type.`
+
+    Parameters
+    ----------
+    lalts : `SwigPyObject`
+        the input LAL time-series
+
+    Returns
+    -------
+    dtype : `type`
+        the numpy data type, such as numpy.uint32, numpy.float64, etc.
+
+    Raises
+    ------
+    ValueError
+        if the input is not a convertible LAL time-series
+    """
+    cls = type(lalts)
+    try:
+        return NUMPY_DTYPE_FROM_LAL_CLASS[cls]
+    except KeyError:
+        raise ValueError(
+            f'Invalid class {cls.__name__!r} for the LAL time-series'
+        ) from None
+
+
```

Comment:
If we need another function to infer types from LAL objects (which you have demonstrated that we do), I would like to try and make it more general, perhaps something like this:

```python
def from_lal_type(laltype):
    if not isinstance(laltype, type):
        laltype = type(laltype)
    name = laltype.__name__
    for nptyp, laltyp in LAL_TYPE_STR_FROM_NUMPY.items():
        if laltyp in name:
            return nptyp
    raise ValueError(
        f"'{name}' has no known numpy type equivalent"
    )
```

I sort of hate this, but I can't find any better way of inferring the LAL type from the LAL object without hardcoding every possible object.

Reply:
I understand the motivation to generalize. I get the idea, but as a note, this won't work for UINT2TimeSeries, the loop will begin with (np.int16, 'INT2') and 'INT2' will match 'UINT2TimeSeries'.
The condition `if name.startswith(laltyp)` should work though but that is still fragile.
Otherwise a regex `(^(U?INT|REAL|COMPLEX)\d+)` to extract the LAL type would be fairly foolproof.

Reply:
Yes, good catch.